server/src/main/java/com/angelomelonas/grpcwebchat/service/ChatSessionService.java

```
@@ -0,0 +1,108 @@
+        ChatSession chatSession = new ChatSession();
+
+        // Add the ChatSession.
+        this.chatSessionHashMap.put(uuid, chatSession);
+
+        final AuthenticationResponse authenticationResponse = AuthenticationResponse.newBuilder()
+                .setUuid(String.valueOf(uuid))
+                .build();
+
+        // Respond with the newly generated UUID.
+        responseObserver.onNext(authenticationResponse);
+        responseObserver.onCompleted();
+
+        LOGGER.info("ChatSession with UUID {} created.", uuid);
+    }
+
+    public void subscribe(UUID uuid, String username, StreamObserver<Message> responseObserver) {
+        ChatSession chatSession = chatSessionHashMap.get(uuid);
+
+        if (chatSession == null) {
+            IllegalArgumentException illegalArgumentException = new IllegalArgumentException("Cannot subscribe. Client not authenticated.");
+            responseObserver.onError(illegalArgumentException);
+            throw illegalArgumentException;
+        }
+
+        chatSession.subscribe(username, responseObserver);
+
+        final Message clientSubscribed = Message.newBuilder()
+                .setUsername("Server")
+                .setMessage("Client subscribed successfully.")
+                .setTimestamp(Instant.now().toEpochMilli())
+                .build();
+
+        responseObserver.onNext(clientSubscribed);
+
+        LOGGER.info("ChatSession with UUID {} and username {} subscribed.", uuid, username);
+    }
+
+    public void unsubscribe(UUID uuid, StreamObserver<Message> responseObserver) {
+        ChatSession chatSession = this.chatSessionHashMap.remove(uuid);
+
+        if (chatSession == null) {
+            IllegalArgumentException illegalArgumentException = new IllegalArgumentException("Cannot unsubscribe. Client not subscribed.");
+            responseObserver.onError(illegalArgumentException);
+            throw illegalArgumentException;
```

do you really want to throw this exception? isn't it good enough to just send the ERROR to the client and then return?